Consolidate and deduplicate MMR API methods

Lederstrumpf opened the pull request:
This PR consolidates the MMR API by merging single-leaf/batch and historical/best-block flavours of proofs, as converged to in #12391. This cuts the API proof methods down to `generate_proof`, `verify_proof`, and `verify_proof_stateless`.

The price paid is that the API is now more verbose for the most simple use cases, in exchange for making the API more deliberate and minimising its surface and maintenance.
Note that the SCALE encoded proofs of single leafs are 2 bytes larger using this universal API (via the singleton vecs `LeavesProof.{leaves, proof.leaf_indices}`). I can't think of a good reason to store these proofs on-chain ever, so I think this has no material impact, but I can add back distinct handling of the single-leaf case if there's good reason.

In my mind, the implication of no longer having a dedicated methods for single-leaf proofs is that the `LeafProof`/`LeavesProof` and `Proof/BatchProof` distinctions are also unnecessary (not consumed elsewhere), so I've merged them too.

Closes #12391.

polkadot companion: https://github.com/paritytech/polkadot/pull/6167

Lederstrumpf commented:
@vgeddes and @seunlanlege, since you're users of this pallet, heads up that merging this PR would break MMR API, both RPC & runtime.

svyatonik reviewed (approved):
LGTM!

acatangiu commented:
@Lederstrumpf did you compile `polkadot` with these changes and verify "generate_proof" and "verify_proof" RPCs are working fine?

directly related: we really need to prioritize https://github.com/paritytech/substrate/issues/11972 so we get some automated testing of this ⬆️

acatangiu commented on frame/merkle-mountain-range/src/lib.rs:
```diff
@@ -382,36 +382,22 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
 
 	/// Generate a MMR proof for the given `block_numbers`.
+	/// If `best_known_block_number = Some(n)`, this generates a historical proof for
+	/// the head of the chain was at height `n`.
```
nit:

```suggestion
	/// If `best_known_block_number = Some(n)`, this generates a historical proof for
	/// the head of the chain at height `n`.
```
or

```suggestion
	/// If `best_known_block_number = Some(n)`, this generates a historical proof for
	/// the chain with head at height `n`.
```

acatangiu reviewed (approved):
code changes look good!

acatangiu commented on frame/merkle-mountain-range/rpc/src/lib.rs:
```diff
@@ -185,6 +210,31 @@ where
+		let decoded_proof =
+			Decode::decode(&mut &proof.proof.0[..]).map_err(runtime_error_into_rpc_error)?;
+
+		api.verify_proof_stateless(
+			&BlockId::hash(proof.block_hash),
+			mmr_root,
+			leaves,
+			decoded_proof,
+		)
+		.map_err(runtime_error_into_rpc_error)?
+		.map_err(mmr_error_into_rpc_error)?;
```
This returns `bool` which is lost - this RPC will return `true` even for invalid proofs.

Lederstrumpf replied:
If the proof is invalid due to incorrect MMR root, then [`verify_leaves_proof`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/mmr/mmr.rs#L33-L37) returns `Ok(false)`, which [`verify_leaves`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/lib.rs#L441-L445) maps to a `Result<(), primitives::Error>`, which is propagated up [here](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/rpc/src/lib.rs#L207-L208). So in this form, the rpc returns `true` only if `verify_leaves_proof` did, and the error returns early otherwise.
If the proof is invalid for other reasons than incorrect MMR root, then [`verify_leaves_proof`](https://github.com/paritytech/substrate/blob/087c77e21a7aad7abbb8d9810989096a8cfd4610/frame/merkle-mountain-range/src/mmr/mmr.rs#L33-L37) returns an error, which is propagated up regardless.

I agree with you though that this is not ideal in terms of type consistency, and this one's unfortunately a tradeoff. These are the other options I considered:

1.  have `verify_leaves` return `RpcResult<()>`. However, the json returns a `null` only then for the unwrapped success case, which in its raw form is misleading to an API consumer.
2.  use a wrapped `RpcResult<Result<(), CallError>>`, which smells antipattern to me
3.  map errors back to `false`, which sacrifices granularity for type consistency

Do you have another preference on how to handle this?

Lederstrumpf commented:
> @Lederstrumpf did you compile `polkadot` with these changes and verify "generate_proof" and "verify_proof" RPCs are working fine?
> 
> directly related: we really need to prioritize #11972 so we get some automated testing of this ⬆️

To follow up on offline discussion we had with @cheme:
`verify_proof` and `verify_proof_stateless` were previously not exposed via RPC, and what we converged on at the retreat was to move these behind `sc_rpc_api::policy::DenyUnsafe`. The motivation of this was to protect consumers of the API against untrusted remote nodes (as they would for proof verification), since `DenyUnsafe` RPCs are unless overridden with `--unsafe-rpc-external` only accessible locally.

Reflecting on this again though: the purpose of `DenyUnsafe` is to protect the node operator against remote nodes (for calls such as `author_insertKey`), and not vice versa, and control for exposing them is with the node operator.
Likewise, I think the trust assumptions on part of the API consumer for MMR proof verification are comparable to querying an untrusted remote node for storage state, which is already exposed via the "safe" RPC API.

So I've added both `mmr_verify_proof` and `mmr_verifyProofStateless` to this PR without `DenyUnsafe` since it now appears consistent to me with existing RPC API practices. Also added `mmr_root` since its return is required for `mmr_verifyProofStateless`.

If there are reservations about exposing these calls via RPC, I can revert the changes or still move the calls behind `DenyUnsafe`.

acatangiu commented on frame/merkle-mountain-range/rpc/src/lib.rs:
```diff
@@ -96,63 +76,59 @@ impl<BlockHash> LeafBatchProof<BlockHash> {
+	fn mmr_root(&self, at: Option<BlockHash>) -> RpcResult<MmrHash>;
 
-	/// Generate a MMR proof for the given `block_numbers` given the `best_known_block_number`.
+	/// Generate an MMR proof for the given `block_numbers`.
```
nit: if you want to go the `an MMR` route, you need to change it everywhere :laughing:

Lederstrumpf replied:
ack 😄. done in https://github.com/paritytech/substrate/pull/12530/commits/212161d5cddf041060025b8ded705cd266c51546.

acatangiu commented:
bot rebase

Lederstrumpf commented on frame/merkle-mountain-range/rpc/src/lib.rs:
```diff
@@ -169,82 +145,98 @@ impl<C, B> Mmr<C, B> {
+		let api = self.client.runtime_api();
+
+		let leaves =
+			Decode::decode(&mut &proof.leaves.0[..]).map_err(runtime_error_into_rpc_error)?;
```
agree `CallError::InvalidParams` is most appropriate. done in https://github.com/paritytech/substrate/pull/12530/commits/63d76941b0b96df31d8c11930a0149fd5d5dda9a.

Lederstrumpf commented on frame/merkle-mountain-range/rpc/src/lib.rs:
```diff
@@ -258,6 +250,21 @@ fn mmr_error_into_rpc_error(err: MmrError) -> CallError {
+			INVALID_BEST_KNOWN_BLOCK_ERROR,
+			"Invalid best known block",
+			Some(data),
+		)),
```
thanks for your detailed take! I've gone with your suggestion to inline using `thiserror` in https://github.com/paritytech/substrate/pull/12530/commits/d5f4cf7de571e11701aeb8738b79df215cd916d4, so error msgs are no longer divorced from the error enum.

> If it makes sense to return CallError::Failed.

We lose error code granularity doing that, and in substrate, `CallError::Failed` is otherwise only used in https://github.com/paritytech/substrate/blob/63d76941b0b96df31d8c11930a0149fd5d5dda9a/client/rpc-api/src/author/error.rs#L179-L180 for handling rpc policy denials, so I've remained with `CallError::Custom`.
The error codes are still defined in `pallet-mmr-rpc`.

serban300 replied:
Looks good. Thanks !

Lederstrumpf commented:
bot rebase
